govcd/disk.go

```diff
@@ -0,0 +1,283 @@
+/*
+ * Copyright 2018 VMware, Inc.  All rights reserved.  Licensed under the Apache v2 License.
+ */
+
+package govcd
+
+import (
+	"bytes"
+	"encoding/xml"
+	"fmt"
+	"github.com/vmware/go-vcloud-director/types/v56"
+	"net/http"
+	"net/url"
+)
+
+// Independent disk
+type Disk struct {
+	Disk   *types.Disk
+	client *Client
+}
+
+// Init independent disk struct
+func NewDisk(cli *Client) *Disk {
+	return &Disk{
+		Disk:   new(types.Disk),
+		client: cli,
+	}
+}
+
+// Create an independent disk in VDC
+// Reference: vCloud API Programming Guide for Service Providers vCloud API 30.0 PDF Page 102 - 103,
+// https://vdc-download.vmware.com/vmwb-repository/dcr-public/1b6cf07d-adb3-4dba-8c47-9c1c92b04857/
+// 241956dd-e128-4fcc-8131-bf66e1edd895/vcloud_sp_api_guide_30_0.pdf
+func (vdc *Vdc) CreateDisk(diskCreateParams *types.DiskCreateParams) (*Disk, error) {
+	var err error
+	var execLink *types.Link
+
+	// Find the proper link for request
+	for _, vdcLink := range vdc.Vdc.Link {
+		if vdcLink.Rel == types.RelAdd && vdcLink.Type == types.MimeDiskCreateParams {
+			execLink = vdcLink
+			break
+		}
+	}
+
+	if execLink == nil {
+		return nil, fmt.Errorf("exec link not found")
+	}
+
+	// Parse request URI
+	reqUrl, err := url.ParseRequestURI(execLink.HREF)
+	if err != nil {
+		return nil, fmt.Errorf("error parse URI: %s", err)
+	}
+
+	// Prepare the request payload
+	diskCreateParams.Xmlns = types.NsVCloud
+
+	xmlPayload, err := xml.Marshal(diskCreateParams)
+	if err != nil {
+		return nil, fmt.Errorf("error xml.Marshal: %s", err)
+	}
+
+	// Send Request
+	req := vdc.client.NewRequest(nil, http.MethodPost, *reqUrl, bytes.NewBufferString(xml.Header+string(xmlPayload)))
+	req.Header.Add("Content-Type", execLink.Type)
+	resp, err := checkResp(vdc.client.Http.Do(req))
+	if err != nil {
+		return nil, fmt.Errorf("error create disk: %s", err)
+	}
+
+	// Decode response
+	disk := NewDisk(vdc.client)
+	if err = decodeBody(resp, disk.Disk); err != nil {
+		return nil, fmt.Errorf("error decoding create disk params response: %s", err)
+	}
+
```

Comment:
Is this very fast operation and we shouldn't return task which return by vCD? It is common practise for function to return tasks and caller can decide to wait or not for result.

Reply:
The tasks is encapsulate in a result of create disk, so it required to retrieve the tasks outside the function like this:

```
task := NewTask(vcd.vdc.client)
	for _, taskItem := range disk.Disk.Tasks.Task {
		task.Task = taskItem
		err = task.WaitTaskCompletion()
		check.Assert(err, IsNil)
	}
```

Do you think there have any better way to return tasks?

Reply:
I think it would be sufficient to use same approach as we use in other places:

`	task := NewTask(orgVdcNet.client)

	if err = decodeBody(resp, task.Task); err != nil {
		return Task{}, fmt.Errorf("error decoding Task response: %s", err)
	}`

Reply:
I think it can do better is ```return (*Disk, []*types.Task ,error)``` because the create disk request will return some new information (e.g. disk href, action href) and the tasks is encapsulate in disk, so return disk and encapsulated tasks to imply there have new disk information and there have new tasks need to be wait.

What do you think? 

Btw nil pointer takes less memory, is there any concern in return a Task instance instead of a nil pointer?

Reply:
At the moment we agreed that our approach is do not return entities due that they return only partial state and thus you will query it anyway. So methods are simple - return task and error. Query when you need entity. In future you want that Task would return entities HREF, but this needs investigation.
We usually return task, not reference to it, that's why it can't be nil.

Reply:
It is updated in new commit.